Add parens to method calls used as arg defaults for Ruby 2.2, fix #3733

rdlugosz opened the pull request:
Ruby 2.2 included a [bug fix](https://bugs.ruby-lang.org/issues/9593) that impacts methods which call other methods as defaults for some of their arguments, e.g., `def bar(foo=foo)`. The fix is to include parenthesis on the method call, `def bar(foo=foo())`. A similar patch was [recently made](https://github.com/rails/rails/commit/e88da370f190cabd1e9750c5b3531735950ab415) to Rails to support Ruby 2.2.

I only found two instances of method calls that used the problematic syntax. Failing tests revealed these methods; the following regex did not find other instances: `def.+\((\S+)\s*=\s*\1\)`.

Additionally, because test-unit is not included with Ruby 2.2 (and is required for ActiveSupport 3.2.x), it has been explicitly included in the Gemfile in the Rails 3.1|2 section.

Tests pass; tested against Rails 3.2.16 & 3-2-stable branch on both Ruby 2.1.5 and 2.2. The JSLint test was not executed because I did not have a JRE set up; I don't think this test is relevant to the issue at hand.

Note that my branch is based off of the 0-6-stable branch, but I suspect this same change is probably relevant to master.

timoschilling commented:
can you please rebase and change "Surrounding space missing in default value assignment."

timoschilling commented:
And can you please change the .travis.yml to test against 2.2

rdlugosz commented:
Hi @timoschilling. Do you mean rebase off of master? I was under the impression that master was intended for Rails 4+, is that not correct? 

I did not include the space in keeping with the existing style (i.e., I added the parens and changed nothing else). Still want spaces?

I can take a look at the Travis config, but I've not used it before. I assume this is a simple change?

timoschilling commented:
I mean a rebase on top of 0-6-stable, the test are failing in your pr while there was a a bug before. 

You only need to change the 2.0 line to 2.2 for Travis.

rdlugosz commented:
Done. However, Travis is now failing on just 2.2+3.0.20; passing for 3.1 and 3.2. I looked at the error log and it's not obvious (to me) why it fails... I *did* have to add the test-unit dep for Rails 3.0 in the Gemfile to get it to build at all.

I'm thinking that I back the Gemfile change out (so as to not change anything related to Rails 3.0 users) and we exclude 2.2+3.0 from the Travis build. Agree? How do we exclude that combination in Travis?

rdlugosz commented:
I have added Ruby 2.2 to the Travis config and `exclude` testing Rails 3.0.x for both Ruby 2.0 and 2.2. It was originally set to `allow_failures` for Ruby 2.0 for all Rails versions, which is a little over-inclusive. Excluding the test altogether is appropriate because Rails 3.0.x is known to not run on Rubies > 1.9.3.

Assuming Travis succeeds, let me know if there's anything else you'd like to see prior to merging this.

timoschilling commented:
good work :+1:

timoschilling commented:
@rdlugosz maybe you find some time to look on rails 3.2 / ruby 2.2 / activeadmin master

rdlugosz commented:
Sure - I'll take a look.

timoschilling commented:
:+1:

rdlugosz commented:
So, I looked into this today and am a little confused because I did *not* find any problems running current master with Rails 4.2 and Ruby 2.2 (aside from the existing, unrelated failing tests in master). I also started up a test 4.2 application**  but did not see any errors related to the file https://github.com/activeadmin/activeadmin/blob/fadbcb8524330ff3d637e2da4f0c4df0f88d213c/lib/active_admin/helpers/collection.rb

I dropped a `raise` in that method to prove that the `collection_size` method is being called (it is), but it seems to work with or without the parens I added to make this work against Rails 3.2/Ruby 2.2. That was unexpected.

I can submit the change to `helpers/collection.rb` to add the parens for Ruby 2.2 but unless I'm missing something I cannot recreate the issue I expected to see. Let me know your thoughts on this.

As for Rails 3.2 support, it looks like the plan is to drop support for anything < Rails 4.1 in the 1.0 release. I don't see any reason to bring back the various conditional statements in the Gemfile if that is the case (they appear to have been removed somewhere between the 0.6.4 branch and master).

-----

** I loosely followed [the directions](https://github.com/activeadmin/activeadmin/blob/master/CONTRIBUTING.md#5-view-your-changes-in-a-rails-application), but they don't quite work—I had to remove the `.test-rails-apps/test-rails-app-4.2.0` and recreate that as a link to `spec/rails/rails-4.2.0` before `script/local server` would give me a working 4.2 app to run locally. Maybe I did something wrong? I can open a separate issue on this unless I'm misunderstanding how that local app is supposed to work.
